equihash/equihash_test.go

```diff
@@ -0,0 +1,844 @@
+	testXorEmptySlice(t, a, b)
+	b, a = a, b
+	testXorEmptySlice(t, a, b)
+	a, b = []byte{}, []byte{}
+	testXorEmptySlice(t, a, b)
+}
+
+func TestXor_NilSlices(t *testing.T) {
+	a := []byte{1, 2, 3, 4, 5}
+	var b []byte
+	testXorEmptySlice(t, a, b)
+	b, a = a, b
+	testXorEmptySlice(t, a, b)
+	var x []byte
+	var y []byte
+	testXorEmptySlice(t, x, y)
+}
+
+func testXor(t *testing.T, a, b, exp []byte) {
+	act := xor(a, b)
+	err := byteSliceEq(act, exp)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestXor_Pass(t *testing.T) {
+	a, b := []byte{0, 1, 0, 1, 0, 1}, []byte{1, 0, 1, 0, 1, 0}
+	exp := []byte{1, 1, 1, 1, 1, 1}
+	testXor(t, a, b, exp)
+	a, b = []byte{1, 0, 1, 0, 1, 0}, []byte{0, 1, 0, 1, 0, 1}
+	testXor(t, a, b, exp)
+	a, b, exp = []byte{0, 0, 1, 1}, []byte{1, 1, 0, 0}, []byte{1, 1, 1, 1}
+	testXor(t, a, b, exp)
+	a, b = []byte{1, 1, 1, 1}, []byte{0, 0, 0, 0}
+	exp = []byte{1, 1, 1, 1}
+	testXor(t, a, b, exp)
+	a, b = b, a
+	testXor(t, a, b, exp)
+}
+
+func TestXor_Fail(t *testing.T) {
+	a, b := []byte{0, 1, 0, 1}, []byte{0, 1, 0, 1}
+	exp := []byte{0, 0, 0, 0}
+	testXor(t, a, b, exp)
+	a, b = []byte{0, 0, 0, 0}, []byte{0, 0, 0, 0}
+	testXor(t, a, b, exp)
+	a, b = []byte{1, 1, 1, 1}, []byte{1, 1, 1, 1}
+	testXor(t, a, b, exp)
+}
+
+func TestHashSize(t *testing.T) {
+	if 64 != hashSize {
+		t.Errorf("hashSize should equal 64 and not %v\n", hashSize)
+	}
+}
+
+type testCountZerosParams struct {
+	h        []byte
+	expected int
+}
+
+func TestCountZeros(t *testing.T) {
+	var paramsSet = []testCountZerosParams{
+		{[]byte{1, 2}, 7},
+		{[]byte{255, 255}, 0},
+		{[]byte{126, 0, 2}, 1},
+		{[]byte{54, 2}, 2},
+		{[]byte{0, 0}, 16},
+	}
+
+	for _, params := range paramsSet {
+		testCountZeros(t, params)
+	}
+}
+
+func testCountZeros(t *testing.T, p testCountZerosParams) {
+	r := countZeros(p.h)
+	if r != p.expected {
+		t.Error("Should be equal, actual:", r, "expected", p.expected)
+	}
+}
+
+func TestJoinBytes(t *testing.T) {
+	a, b := []byte{1, 2, 3, 4}, []byte{5, 6, 7, 8}
+	act, exp := joinBytes(a, b), []byte{1, 2, 3, 4, 5, 6, 7, 8}
+	err := byteSliceEq(act, exp)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func solutionEq(x, y []int) error {
+	return intSliceEq(x, y)
+}
+
+func TestExccPerson_2(t *testing.T) {
+	p := exccPerson(N, K)
+	n := len(exccPrefix)
+	err := byteSliceEq(p[:n], []byte(exccPrefix))
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+	err = byteSliceEq(p[n:n+4], writeU32(uint32(N)))
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+	err = byteSliceEq(p[n+4:n+8], writeU32(uint32(K)))
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+}
+
+func TestWriteU32(t *testing.T) {
+	exp := []byte{1, 0, 0, 0}
+	err := byteSliceEq(writeU32(1), exp)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestPutU32(t *testing.T) {
+	exp, act := []byte{1, 0, 0, 0}, make([]byte, 4)
+	putU32(act, 1)
+	err := byteSliceEq(act, exp)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func testHashKeys(n int) []hashKey {
```

Comment:
`testHashKeys` function looks like a test case. I'd suggest naming it `generateHashKeys`